Patched Bug #63

kb1lqd opened the pull request:
Fixed #63 and updated other similar items.

* Added length test status and check for global failure to deviceconfig.py so that data too long errors (could update to truncation later if wanted)
* Added other generic length checks for good practice
* Added custom padding packet creation function in commandmodule.py to pad data with specific bytes
* Force all Lat, Lon, Alt to floats and then convert back to string (all NMEA strings contain decimals)
* Padding function on Lat, Lon, Alt add 0x30 (ASCII 0) to each unused byte after data. This matches intended GPS NMEA DMM format in ASCII.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into Firmware_Revision_Telem

Commit:
Basic checks updates in device configuration files.

Commit:
Adding temp proxy settings

Commit:
Correctly checking and erroring on config to long lengths

Added global failure that prevents programming from occuring if a length
is too long.,

Commit:
Completed initial effort

Force all GPS lat/lon/alt to float and also create packet with ASCII
padding (0x30 for 0's) just like NMEA ASCII uses.

Commit:
Removed prints and added comments.

Commit:
Generic configuration

Update fields to generic configurations.

Commit:
Generic.

Commit:
New scripts creating memory checks

Commit:
Cleaned version creating TI script

Commit:
Com port variable

Working com port variable for scripter.

Commit:
Added cmd ARGUMENTS

Commit:
Test programming

Test.py is programming from command line!

Commit:
Deleted Old Files

Commit:
Cleaned a bit

Commit:
Firmware Folder

Commit:
Cleaned up single BAT

Single BAT file for running working. Not as cleanly modular as I'd like
to much better than before.

Commit:
Finalized

Finished the example and tutorial first pass.

Commit:
Fixed README images

Fixed absolute links that were broken to relative for GitHub.

Commit:
Basic folder structure cleaning

Deleted and renamed folders to better organize.

Commit:
Generic proxy settins tutorial

Changed to REPLACEME's

local_device_callsign = 'REPLACEME'  # Should match the connected
Faraday unit as assigned in Proxy configuration
local_device_node_id = REPLACEME  # Should match the connected Faraday
unit as assigned in Proxy configuration

Commit:
Fixed Tutorial 0

Commit:
Tutorial 0 Generic

Proxy config

Commit:
Cleaned Tutorial 2

Just inserted a few newlines to clean up output.

kb1lqc committed:
KB1LQD stop commiting your callsign!

kb1lqc commented on Applications/Telemetry/telemetry.ini:
```diff
@@ -9,4 +9,4 @@ schemaname=db.sql
 [telemetry]
 units=1
 unit0call=REPLACEME
-unit0id=REPLACEME
```
I removed your callsign

kb1lqc committed:
Removed callsign

kb1lqc committed:
Removed callsign

kb1lqc committed:
Removed callsign and com port

kb1lqc commented on Applications/deviceconfiguration/faraday_config.ini:
```diff
@@ -12,14 +12,14 @@ BOOT_FREQUENCY_MHZ = 914.5
 BOOT_RF_POWER = 20
 
 [GPS]
-DEFAULT_LATITUDE = 012345678
```
You completely deleted DEFAULT_LATITUDE... make sure to test code prior to pull request...

kb1lqc commented on Applications/deviceconfiguration/faraday_config.ini:
```diff
@@ -12,14 +12,14 @@ BOOT_FREQUENCY_MHZ = 914.5
-DEFAULT_ALTITUDE = 0
-DEFAULT_ALTITUDE_UNITS = M
+;NMEA DMM Format
+DEFAULT_LATITUDE_DIRECTION = N ; MAX LENGTH = 1 Byte
```
While inline comments may have worked for you @kb1lqd they are not the standard from Windows which defined the INI file structure and therefore are not guaranteed to work.

https://stackoverflow.com/questions/1378219/do-standard-windows-ini-files-allow-comments#19550081

kb1lqc committed:
Removed inline comments and added default latitude

kb1lqc committed:
Removed a stray inline comment

kb1lqc reviewed (dismissed):
I already changed a bunch of code thinking it would be simple fixes. However, When programming GPS lat/long I noticed that leading zeros were removed. You probably need to not convert to floats or at least directly (test then use string). 

This is an example APRS string which uses NMEA GPS data. I has leading zeros. You need to support this or we need to justify why not and pad zeros on the front in APRS only.

APRS,TCPXX*,qAX,CWOP-7:@130726z3729.30N/07959.48W_322/006g016t041r000p000P000b10160h40eMB32

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into Developer_Tutorial_Clean

kb1lqd commented:
@kb1lqc Yes you are right that the parsing on the RECEIVE side is dropping the leading zero(s) and this is unwanted. This was unintentional.

![image](https://cloud.githubusercontent.com/assets/339144/22916698/cfdafc3a-f234-11e6-9e77-d15e5caa5387.png)

Printing the latitude string prior and post the conversion shows the obvious results.
![image](https://cloud.githubusercontent.com/assets/339144/22916680/b920a7b0-f234-11e6-94c7-756e182ad057.png)


I removed the float() and left the pre-packing input to a string of ASCII bytes and it worked except for altitude:

![image](https://cloud.githubusercontent.com/assets/339144/22917244/532f3b70-f238-11e6-8946-ba200658c9bc.png)


![image](https://cloud.githubusercontent.com/assets/339144/22917232/47f03f7a-f238-11e6-9a64-60d0a66dcace.png)

@kb1lqc Looks like the actual problem remaining in Telemetry.py for the Altitude being not text but a "Real" (integer?).

![image](https://cloud.githubusercontent.com/assets/339144/22917130/c0f0a58c-f237-11e6-8bf7-119f2541d128.png)

kb1lqd commented:
OK @kb1lqc I updated the functions to PREPEND instead of APPEND the ASCII zeros:

![image](https://cloud.githubusercontent.com/assets/339144/22917626/be10eefa-f23a-11e6-9af2-624e1646992e.png)


This makes sense from an NMEA formating perspective and also allows the use of INT instead of Float for Alt (i.e. 185 instead of forcing 185.0). In the end I think we'll be OK with this.

NOTE: Should we update the Telemetry.py SQL field for ALT from REAL to TEXT to display all characters (dropped zeros) as they are in the data going to the SQL databate?

Commit:
Fixed Bryces comments

https://github.com/FaradayRF/Faraday-Software/pull/68#pullrequestreview-21449392

* Prepend instead of append Zero's
* Do not convert to float, all Lat, Lon, Alt remains ASCII string
* Fixed bug in the submitted deviceconfig.py that padded ALTITUDE with
the ALTITUDE_UNIT length (1 instead of 8 bytes)

kb1lqd commented:
I changed (did not commit) the telemetry SQL format for ALTITUDE to TEXT instead of real and it changed as expected:

![image](https://cloud.githubusercontent.com/assets/339144/22917760/932a309c-f23b-11e6-896e-e5b252ee8a38.png)

@kb1lqc Do you want to save ALT as a text string like so or INT as before?

kb1lqd commented:
P.S. A nifty GPS converter tool for NMEA DMM to other formats located here: http://www.csgnetwork.com/gpscoordconv.html

Commit:
Updating with formatting check

kb1lqd commented:
NMEA Format: http://navspark.mybigcommerce.com/content/NMEA_Format_v0.1.pdf

![image](https://cloud.githubusercontent.com/assets/339144/22964380/2dc72b42-f30d-11e6-8c09-574aebf55a29.png)

Updates:
* I format GPS flash configuration data to be in formats
  * Lat: ddmm.mmmm (leading and trailing 0x30's ('0') padding bytes added automatically)
  * Lon: dddmm.mmmm (leading and trailing 0x30's ('0') padding bytes added automatically)
  * Alt: 8 Bytes forced to a Float (leading 0x30's ('0') padding bytes added automatically)

Note that _I deviate from the linked Specifical by also adding trailing pad bytes to Lat and Lon but justify this with clean userspace if only generic location is wanted._

Commit:
Format GPS default flash NMEA

GGA format.

Commit:
Cleaned

Removed some engineering data prints.

Commit:
GPS checking format working

Working!

kb1lqd commented:
Checking that input performs correct formatting:

## Correct Size Lat/Lon/Alt

![image](https://cloud.githubusercontent.com/assets/339144/22965262/997635fa-f311-11e6-957e-5d54bd26b710.png)


![image](https://cloud.githubusercontent.com/assets/339144/22965260/95b873b0-f311-11e6-860b-3de8afe47316.png)


## Short Lat/Lon/Alt:

Test cases worked.

![image](https://cloud.githubusercontent.com/assets/339144/22965162/109aeb90-f311-11e6-9f0b-6880454b0b17.png)

![image](https://cloud.githubusercontent.com/assets/339144/22965160/08f5301c-f311-11e6-9b25-b8fb8b99d8b6.png)

## Long Lat/Lon/Alt

Correctly errors.

## Incorrect Format (no decimal)

![image](https://cloud.githubusercontent.com/assets/339144/22965402/472c1b06-f312-11e6-90eb-c02daa54f255.png)

Correctly failed

![image](https://cloud.githubusercontent.com/assets/339144/22965415/5861c330-f312-11e6-96ac-0afdafc8d30e.png)

# Altitude Check

* Checks correctly for maximum of< 18km
* I chose to NOT support negative altitude at this time. I couldn't think of a nice way to format this in NMEA correctly and decided to move on. Sorry Death Valley, CA residents!
* Correctly checks length in bytes, converters to float as needed

Commit:
All checks for GPS formatting work

As intended!

Commit:
Replaced with generic configuration

Commit:
Generic config

kb1lqd commented:
@kb1lqc Please review again for implemented updates.

kb1lqc commented:
@kb1lqd did you commit? My Git log shows no new changes to New_Bug_63 have been made... Oh wait you branched again to Developer_Tutorial_Clean. Try to keep the branches the same. I'll have to merge Developer_Tutorial_Clean with this branch to keep the pull request valid... though it may pull in your tutorial updates..

kb1lqd committed:
Merge branch 'New_Bug_63' into Developer_Tutorial_Clean

kb1lqd committed:
Merge pull request #72 from FaradayRF/Developer_Tutorial_Clean

Developer tutorial clean Into New Bug 63

kb1lqd committed:
Revert "Developer tutorial clean Into New Bug 63"

kb1lqd committed:
Merge pull request #73 from FaradayRF/revert-72-Developer_Tutorial_Clean

Revert "Developer tutorial clean Into New Bug 63"

kb1lqd commented:
Moved this merge over to #76
